### PR TITLE
 Refactored yii\base\Module. Added tests

### PR DESCRIPTION
**Pull request by vladis84**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes/no
| Tests pass?   | yes

### CODE REVIEW

File: framework/base/Module.php [hunk 1]
```diff
@@ -705,4 +705,60 @@ public function afterAction($action, $result)
         $this->trigger(self::EVENT_AFTER_ACTION, $event);
         return $event->result;
     }
+
+    /**
+     * Layout find in parent module.
+     * @return array [module, layout]
+     */
+    private function findLayout()
```
Comment (anton-pribora) [inline framework/base/Module.php]: A little bit better I think
```php
private function findLayout()
{
    $module  = $this;

    do {
        if (is_string($module->layout)) {
            return [$module, $module->layout];
        }
    } while (($module = $module->module) !== null);

    return [null, null];
}
```

File: framework/base/Module.php [hunk 2]
```diff
@@ -705,4 +705,60 @@ public function afterAction($action, $result)
...
+        }
+
+        if (!is_string($layout)) {
+            return false;
+        }
+
+        if (Yii::isAlias($layout)) {
+            $file = Yii::getAlias($layout);
+        } else {
+            $file = $module->getLayoutPath() . DIRECTORY_SEPARATOR . ltrim($layout, '/');
```
Comment (anton-pribora) [inline framework/base/Module.php]: I think it should be one of `DIRECTORY_SEPARATOR . ltrim($layout, DIRECTORY_SEPARATOR);` or `'/' . ltrim($layout, '/')`

**Review by anton-pribora — APPROVED** [review]

File: framework/base/Controller.php [hunk 3]
```diff
@@ -479,40 +479,10 @@ public function setViewPath($path)
      * Please refer to [[render()]] on how to specify this parameter.
      * @throws InvalidParamException if an invalid path alias is used to specify the layout.
      */
-    public function findLayoutFile($view)
+    public function findLayoutFile(View $view)
```
Comment (samdark) [inline framework/base/Controller.php]: Method signature can't be changed in Yii 2.0.x because of backwards compatibility.

Comment (vladis84) [inline framework/base/Controller.php]: ok

File: framework/BaseYii.php [hunk 4]
```diff
@@ -128,7 +128,7 @@ public static function getVersion()
      */
     public static function getAlias($alias, $throwException = true)
     {
-        if (strncmp($alias, '@', 1)) {
+        if (!self::isAlias($alias)) {
```
Comment (cebe) [inline framework/BaseYii.php]: wrapping a function call in an extra method is unnecessary.

File: framework/BaseYii.php [hunk 5]
```diff
@@ -156,6 +156,15 @@ public static function getAlias($alias, $throwException = true)
     }
 
     /**
+     * @param string $alias
+     * @return bool
+     */
+    public static function isAlias($alias)
```
Comment (cebe) [inline framework/BaseYii.php]: what is the use case for this method? There should not be a need to check for this.

**Review by cebe — COMMENTED** [review]
can you explain a bit why you made these changes? i.e. move the logic from controller to module.

File: framework/BaseYii.php (hunk 5 above)
Comment (vladis84) [inline framework/BaseYii.php]: https://github.com/yiisoft/yii2/pull/14085/files#diff-d24221cf7f251030273dc9f1c436ba75R747
Code reuse

**Comment by vladis84** [comment]
I think it is the duty of the module

File: framework/BaseYii.php (hunk 4 above)
Comment (vladis84) [inline framework/BaseYii.php]: Public api will be broken

Comment (cebe) [inline framework/BaseYii.php]: how exactly? This adds unnecessary overhead.

Comment (vladis84) [inline framework/BaseYii.php]: This method return 3 value.
First original parameter.
To alias path.
Three value of of alias not found false

File: framework/base/Module.php [hunk 6]
```diff
@@ -705,4 +705,58 @@ public function afterAction($action, $result)
...
+
+    /**
+     * Layout find in module.
+     * @param string $layout    Layout name
+     * @param string $layoutExt View default extension
+     * @return boolean|string
+     */
+    public function findLayoutFile($layout, $layoutExt)
+    {
+        $module = $this;
```
Comment (dynasource) [inline framework/base/Module.php]: $module is confusing here. This should be in the next `else` statement

File: framework/base/Module.php [hunk 7]
```diff
@@ -705,4 +705,58 @@ public function afterAction($action, $result)
...
+            }
+        }
+        while (($module = $module->module) !== null);
+
+        return [null, null];
+    }
+
+    /**
+     * Layout find in module.
+     * @param string $layout    Layout name
```
Comment (dynasource) [inline framework/base/Module.php]: this should be called $controllerLayout or $defaultLayout for a better understanding

File: framework/base/Module.php (hunk 1 above)
Comment (dynasource) [inline framework/base/Module.php]: this method feels awkward. Why not simply use
```php
private function findLayout()
{
    $module = $this;
    while(module = $module->module){
        if (is_string($module->layout)) {
            return [$module, $module->layout];
        }
    }
    return false;
}
```

Comment (dynasource) [inline framework/base/Module.php]: additionally, it returns the $module + $layout, so the method should match this return result

**Review by dynasource — CHANGES_REQUESTED** [review]

**Comment by samdark** [comment]
@vladis84 are you going to work more on it?

**Comment by samdark** [comment]
Won't be able to merge this into 2.0 and for 2.1 it doesn't make sense. Sorry.

**Pull request closed without merging**